src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs

```diff
@@ -227,6 +449,95 @@ public void Dispose()
             }
         }
 
+        private static bool CompareTags(IEnumerable<KeyValuePair<string, object?>>? tags1, IEnumerable<KeyValuePair<string, object?>>? tags2)
+        {
+            if (tags1 is null)
+            {
+                return tags2 is null;
+            }
+
+            if (tags2 is null)
+            {
+                return false;
+            }
+
+            if (tags1 is ICollection<KeyValuePair<string, object?>> firstCol && tags2 is ICollection<KeyValuePair<string, object?>> secondCol)
+            {
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IList<KeyValuePair<string, object?>> firstList && secondCol is IList<KeyValuePair<string, object?>> secondList)
+                {
+                    int count = firstList.Count;
+                    for (int i = 0; i < count; i++)
+                    {
+                        KeyValuePair<string, object?> pair1 = firstList[i];
+                        KeyValuePair<string, object?> pair2 = secondList[i];
+                        if (pair1.Key != pair2.Key || !object.Equals(pair1.Value, pair2.Value))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            using (IEnumerator<KeyValuePair<string, object?>> e1 = tags1.GetEnumerator())
+            using (IEnumerator<KeyValuePair<string, object?>> e2 = tags2.GetEnumerator())
+            {
+                while (e1.MoveNext())
+                {
+                    KeyValuePair<string, object?> pair1 = e1.Current;
+                    if (!e2.MoveNext() || pair1.Key != e2.Current.Key || !object.Equals(pair1.Value, e2.Current.Value))
+                    {
+                        return false;
+                    }
+                }
+
+                return !e2.MoveNext();
+            }
+        }
+
+        // AddInstrument will be called when publishing the instrument (i.e. calling Instrument.Publish()).
```

Comment:
Is "AddInstrument" here referring to "GetOrCreateInstrument"?  I'm not sure why this comment is here otherwise.

Reply:
Here are how the calls go:

`GetOrCreateInstrument`  -> `instrumentCreator.Invoke();` -> `Instrument.Publish` -> `AddInstrument`